Realm/RLMObservation.hpp

```diff
@@ -0,0 +1,143 @@
+    void didChange(NSString *key, NSKeyValueChange kind=NSKeyValueChangeSetting, NSIndexSet *indexes=nil) const;
+
+    bool isForRow(size_t ndx) const {
+        return row && row.get_index() == ndx;
+    }
+
+    void recordObserver(realm::Row& row, RLMObjectSchema *objectSchema, NSString *keyPath);
+    void removeObserver();
+    bool hasObservers() const { return observerCount > 0; }
+
+    // valueForKey: on observed object and array properties needs to return the
+    // same object each time for KVO to work at all. Doing this all the time
+    // requires some odd semantics to avoid reference cycles, so instead we do
+    // it only to the extent specifically required by KVO. In addition, we
+    // need to continue to return the same object even if this row is deleted,
+    // or deleting an object with active observers will explode horribly.
+    // Once prepareForInvalidation() is called, valueForKey() will always return
+    // the cached value for object and array properties without checking the
+    // backing row to verify it's up-to-date.
+    id valueForKey(NSString *key);
+
+    void prepareForInvalidation();
+
+private:
+    // Doubly-linked-list of observed objects for the same row as this
+    RLMObservationInfo *next = nullptr;
+    RLMObservationInfo *prev = nullptr;
+
+    // Row being observed
+    realm::Row row;
+    RLMObjectSchema *objectSchema;
+
+    // Object doing the observing
+    __unsafe_unretained id object;
+
+    // valueForKey: hack
+    bool invalidated = false;
+    size_t observerCount = 0;
+    NSString *lastKey = nil;
+    __unsafe_unretained RLMProperty *lastProp = nil;
+
+    // objects returned from valueForKey() to keep them alive in case observers
+    // are added and so that they can still be accessed after row is detached
+    NSMutableDictionary *cachedObjects;
+
+    void setRow(realm::Table &table, size_t newRow);
+
+    template<typename F>
+    void forEach(F&& f) const {
+        for (auto info = prev; info; info = info->prev)
+            f(info->object);
+        for (auto info = this; info; info = info->next)
+            f(info->object);
+    }
+
+    // Default move/copy constructors don't work due to the intrusive linked
+    // list and we don't need them
+    RLMObservationInfo(RLMObservationInfo const&) = delete;
+    RLMObservationInfo(RLMObservationInfo&&) = delete;
+    RLMObservationInfo& operator=(RLMObservationInfo const&) = delete;
+    RLMObservationInfo& operator=(RLMObservationInfo&&) = delete;
+
+public:
+    // storage for the observationInfo property on RLMObjectBase
+    void *kvoInfo = nullptr;
+};
+
+// Get the the observation info chain for the given row
+// Will simply return info if it's non-null, and will search ojectSchema's array
+// for a matching one otherwise, and return null if there are none
+RLMObservationInfo *RLMGetObservationInfo(std::unique_ptr<RLMObservationInfo> const& info,
```

Comment:
Why take a `const std::unique_ptr&` rather than taking a raw pointer? Does `RLMGetObservationInfo` for some reason care that the caller owns the `RLMObservationInfo` that's being passed as an argument?